common/internal_model/src/main/scala/uk/ac/wellcome/models/work/internal/AbstractRootConcept.scala

```diff
@@ -15,6 +21,57 @@ object Concept {
   }
 }
 
+case class DateRange(label: String,
+                     start: LocalDateTime,
+                     end: LocalDateTime,
+                     inferred: Boolean)
+    extends AbstractConcept
+object DateRange {
+  type GetDateRange = (String, LocalDateTime) => DateRange
+  type DateRangeParser = (DateTimeFormatter, GetDateRange)
+
+  val parsers: List[(String, GetDateRange)] = List(
+    (
+      "yyyy",
+      (label: String, start: LocalDateTime) =>
+        DateRange(
+          label,
+          start,
+          end = start.plusYears(1).minusNanos(1),
+          inferred = false)
+    ),
+    (
+      "'['yyyy']'",
+      (label: String, start: LocalDateTime) =>
+        DateRange(
+          label,
+          start,
+          end = start.plusYears(1).minusNanos(1),
+          inferred = true)
+    )
+  )
+
+  private def formatterWithDefaults(pattern: String): DateTimeFormatter =
+    new DateTimeFormatterBuilder()
+      .appendPattern(pattern)
+      .parseDefaulting(ChronoField.NANO_OF_SECOND, 0)
+      .parseDefaulting(ChronoField.SECOND_OF_MINUTE, 0)
+      .parseDefaulting(ChronoField.MINUTE_OF_HOUR, 0)
+      .parseDefaulting(ChronoField.HOUR_OF_DAY, 0)
+      .parseDefaulting(ChronoField.DAY_OF_MONTH, 1)
+      .parseDefaulting(ChronoField.MONTH_OF_YEAR, 1)
+      .parseDefaulting(ChronoField.YEAR_OF_ERA, Year.now().getValue())
+      .toFormatter()
+
+  def parse(label: String): Option[DateRange] = {
+    parsers map {
+      case (pattern, getDateRange) =>
+        Try(LocalDateTime.parse(label, formatterWithDefaults(pattern)))
+          .map(getDateRange(label, _))
+    } find (_.isSuccess) map (_.get)
```

Comment:
This is all a bit mushed together – it looks like it goes through the list, then calls `formatterWithDefaults` (what does that do?), and then does some `.isSuccess` and `.get`. Could we break it up to make it easier to read?

Reply:
Yea, I might take a look at turning into a `tailref`, I don't really like this.